main/src/library/Fixpoint3/Phase/Lowering.flix

```diff
@@ -0,0 +1,585 @@
+/*
+ * Copyright 2025 Casper Dalgaard Nielsen
+ *                Adam Yasser Tallouzi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+///
+/// This phase transforms the abstract `Ram` into a low-level `ExecutableRam` in a format
+/// the Interpreter understands.
+///
+/// Concretely it does the following
+///
+/// 1. Replaces `Boxed` by `Int64`, except for lattice-elements
+/// 2. For relations with multiple indexes replace `MergeInto(newRel, otherRel)` with
+/// `MergeInto(newRel, otherRelIndex1)`, `MergeInto(newRel, otherRelIndex2)`, ...
+/// 3. Does similar for `Purge` and `Swap`.
+/// 4. Augment every statement with information about where the relevant
+/// indexes are stored
+/// 5. Remove `NotBot` as it is handled explicitly by the interpreter.
+/// 6. Assign memory positions to `RowVar`'s.
+/// 7. Collect information about where `RowVar` `rv1`'s values are needed.
+///
+/// An example of (7) would be the following transformation from
+/// ```
+/// search a ∈ A do
+///     query {b ∈ B | b[0] = a[0]} do
+///         project (a[0]) into R
+///     end
+/// end
+/// ```
+/// to
+/// ```
+/// search a ∈ A | write a[0] lookupArrForB[0] do
+///     query {b ∈ B | lookupArrForB } do
+///         project (a[0]) into R
+///     end
+/// end
+/// ```
+///
+/// The purpose is to only do this once and avoid creating new objects constantly.
+/// If the nesting was deeper we only do the write once when we bind `a`, not every
+/// time we have to query `b`. We also only create `lookupArrForB` once, if we ignore
+/// parallelism.
```

Comment:
I assume this related to the above example, but I don't understand the text at all.